File: 146. LRU Cache.js
var linkedNode = function(key, value) {
    this.key = key;
    this.value = value;
    this.pre = this.next = null;
};
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.capacity = capacity;
    this.hash = new Map();
    this.head = null;
    this.tail = null;
};

/**
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if (this.hash.has(key)) {
        var node = this.hash.get(key);
        this.remove(node);
        this.setHead(node);
        return node.value;
    } else {
        return -1;
    }
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if (this.hash.has(key)) {
        var oldNode = this.hash.get(key);
        oldNode.value = value
        this.remove(oldNode);
        this.setHead(oldNode);
    } else {
        var newNode = new linkedNode(key, value);
        if (this.hash.size >= this.capacity) {
            this.hash.delete(this.tail.key);
            this.remove(this.tail);
            this.setHead(newNode);
        } else {
            this.setHead(newNode);
        }
        this.hash.set(key, newNode);
    }
};

LRUCache.prototype.remove = function(node) {
    if (node.pre) {
        node.pre.next = node.next;
    } else {
        this.head = node.next;
    }

    if (node.next) {
        node.next.pre = node.pre;
    } else {
        this.tail = node.pre;
    }
};

LRUCache.prototype.setHead = function(node) {
    node.next = this.head;
    node.pre = null;

    if (this.head) {
        this.head.pre = node;
    }

    this.head = node;

    if (!this.tail) {
        this.tail = this.head;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = Object.create(LRUCache).createNew(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */

File: 146. LRU Cache.py
class LinkedNode:
    
    def __init__(self, key=None, value=None, pre = None, next=None):
        self.key = key
        self.value = value
        self.pre = pre
        self.next = next
        
class LRUCache:
    
    def __init__(self, capacity):
        self.hash = {}
        self.head = None
        self.tail = self.head
        self.capacity = capacity

    def get(self, key):
        if key not in self.hash:
            return -1
        else:
            n = self.hash[key]
            self.remove(n)
            self.setHead(n)
            return n.value

    def put(self, key, value):
        if key in self.hash:
            n = self.hash[key]
            n.value = value
            self.remove(n)
            self.setHead(n)
        else:
            n = LinkedNode(key, value)
            if len(self.hash) >= self.capacity:
                del(self.hash[self.tail.key])
                self.remove(self.tail)
                self.setHead(n)
            else:
                self.setHead(n)
            self.hash[key] = n
    
    def remove(self, n):
        if n.pre is not None:
            n.pre.next = n.next
        else:
            self.head = n.next
        
        if n.next is not None:
            n.next.pre = n.pre
        else:
            self.tail = n.pre
    
    def setHead(self, n):
        n.next = self.head
        n.pre = None
        
        if self.head is not None:
            self.head.pre = n
        
        self.head = n
        
        if self.tail is None:
            self.tail = self.head
            

